src/main/java/com/bootcamp/demo/dao/FirestoreDaoImpl.java

```diff
@@ -74,4 +72,53 @@ public void update(String id, String collectionName, Map cardDetails) {
         }
         return cards;
     }
+
+    @Override
+    public void addCard(Card card) {
+        ApiFuture<QuerySnapshot> findAllQuery = firestoreDB.collection(CARDS).get();
+        try {
+        Map<String, Object> cardData = new HashMap<>();
+        cardData.put("cardNumber", card.getCardNumber());
+        cardData.put("holderName", card.getHolderName());
+        cardData.put("CVV", card.getCVV());
+        cardData.put("state", findAllQuery.get().getDocuments().isEmpty()
+                ? CardStateEnum.PREFERRED
+                : CardStateEnum.NONE);
+        cardData.put("expirationYear", card.getExpirationYear());
+        cardData.put("expirationMonth", card.getExpirationMonth());
+
+        ApiFuture<WriteResult> document = firestoreDB.collection(CARDS).document().set(cardData);
+
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
```

Comment:
If you want you could log this exception: `LOGGER.error("Failed to add a card", e);`
The logger should be a field in this class: `private static final Logger LOGGER = LoggerFactory.getLogger(FirestoreDaoImpl.class);`

Reply:
Note: if you just catch the exception and not throw it to the Service and Controller levels the user (REST client) will not get any kind of feedback when the add operation fails.
If you want you could catch/handle the exception and return an appropriate response for this method - e.g. false in case of exception